vault/expiration.go

```diff
@@ -245,132 +290,94 @@ func (m *ExpirationManager) Restore() error {
-	// Buffer these channels to prevent deadlocks
-	errs := make(chan error, len(existing))
-	result := make(chan *leaseEntry, len(existing))
-
-	// Use a wait group
-	wg := &sync.WaitGroup{}
-
-	// Create 64 workers to distribute work to
-	for i := 0; i < consts.ExpirationRestoreWorkerCount; i++ {
-		wg.Add(1)
-		go func() {
-			defer wg.Done()
-
-			for {
-				select {
-				case leaseID, ok := <-broker:
-					// broker has been closed, we are done
-					if !ok {
-						return
-					}
-
-					le, err := m.loadEntry(leaseID)
-					if err != nil {
-						errs <- err
-						continue
-					}
-
-					// Write results out to the result channel
-					result <- le
-
-				// quit early
-				case <-quit:
-					return
-				}
-			}
-		}()
-	}
-
-	// Distribute the collected keys to the workers in a go routine
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-		for i, leaseID := range existing {
-			if i%500 == 0 {
-				m.logger.Trace("expiration: leases loading", "progress", i)
-			}
-
-			select {
-			case <-quit:
-				return
-
-			default:
-				broker <- leaseID
-			}
-		}
-
-		// Close the broker, causing worker routines to exit
-		close(broker)
-	}()
-
 	// Restore each key by pulling from the result chan
-	for i := 0; i < len(existing); i++ {
-		select {
-		case err := <-errs:
-			// Close all go routines
-			close(quit)
+	restoredCount := 0
+	for i, leaseID := range existing {
+		if i > 0 && i%500 == 0 {
+			m.logger.Trace("expiration: leases loading", "progress", i)
+		}
 
+		le, err := m.loadAndRestoreLease(leaseID)
+		if err != nil {
 			return err
+		}
+		if le != nil {
+			restoredCount++
+		}
 
-		case le := <-result:
+		// This exists for test only to add latency to load calls
+		if loadDelay > 0 {
+			time.Sleep(loadDelay)
+		}
+	}
 
-			// If there is no entry, nothing to restore
-			if le == nil {
-				continue
-			}
+	// Clear our the restored entries and turn off restore mode
+	m.restoreMutex.Lock()
+	m.restoreLoaded = make(map[string]struct{})
+	atomic.StoreInt64(&m.restoreMode, 0)
+	m.restoreMutex.Unlock()
 
-			// If there is no expiry time, don't do anything
-			if le.ExpireTime.IsZero() {
-				continue
-			}
+	m.logger.Info("expiration: leases restored", "restored_lease_count", restoredCount)
+	return nil
+}
 
-			// Determine the remaining time to expiration
-			expires := le.ExpireTime.Sub(time.Now())
-			if expires <= 0 {
-				expires = minRevokeDelay
-			}
+func (m *ExpirationManager) loadAndRestoreLease(leaseID string) (*leaseEntry, error) {
+	m.restoreMutex.Lock()
+	defer m.restoreMutex.Unlock()
 
-			// Setup revocation timer
-			m.pending[le.LeaseID] = time.AfterFunc(expires, func() {
-				m.expireID(le.LeaseID)
-			})
-		}
+	// If we loaded a lease elsewhere in the expiration manager while in
+	// restore, skip the entry since it may have been updated since we
+	// loaded it
+	if _, ok := m.restoreLoaded[leaseID]; ok {
+		return nil, nil
 	}
 
-	// Let all go routines finish
-	wg.Wait()
+	le, err := m.loadEntryOnly(leaseID)
```

Comment:
Not sure we need the two functions anymore? In both cases i believe the final logic is the same:
```
loadLease()
if inRestoremode {
    restoreLease()
}
```